src/mscorlib/src/System/Diagnostics/Eventing/EventSource.cs

```diff
@@ -5803,6 +5811,62 @@ public enum EventManifestOptions
         /// </summary>
         AllowEventSourceOverride = 0x8,
     }
+#if FEATURE_CORECLR && FEATURE_PAL   
+    internal  class EventLogger : EventListener
+    {
+        public EventLogger() {  }
+
+        private static string Serialize(ReadOnlyCollection<string> payloadname, ReadOnlyCollection<object> payload, string sep = ", ")
+        {
+
+            StringBuilder res = new StringBuilder();
+            var fielditer = payloadname.GetEnumerator();
+            var valueiter = payload.GetEnumerator();
+
+            while ((fielditer.MoveNext()) && (valueiter.MoveNext()))
+            {
+                var field = fielditer.Current.ToString();
+
+                var valuestr = valueiter.Current.ToString();
+                res.Append(field);
+                res.Append(":");
+                res.Append(valuestr);
+                res.Append(sep);
+
+            }
+            return res.ToString(0,res.Length - sep.Length);
+
+        }
+
+        internal protected  override void OnEventSourceCreated(EventSource eventSource)
+        {
+            EnableEvents(eventSource, EventLevel.LogAlways, EventKeywords.All, null);
+        }
+
+
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        private static extern void LogEventSource(int eventID, string eventName, string eventSourceName, string payload);
```

Comment:
I take it that this EventSource code will never be part of Microsoft.Diagnostics.Tracing since that cannot depend upon an CoreCLR QCall. Correct?